polynote-kernel/src/main/scala/polynote/kernel/context/RuntimeContext.scala

```diff
@@ -0,0 +1,118 @@
+package polynote.kernel.context
+
+import java.util.concurrent.ConcurrentHashMap
+
+import scala.collection.immutable.Queue
+import scala.collection.{TraversableLike, immutable, mutable}
+import scala.reflect.ClassTag
+
+/**
+  * RuntimeContext keeps track of the results of running each cell, and enforces reproducibility by specifying the
+  * 'state of the world' for each cell of a notebook based on its position in the notebook relative to its run order.
+  *
+  * When each cell is run it is provided with a RuntimeContextView of the proper state in that location, where the proper
+  * state in this case refers to the order of the cells in the notebook.
+  *
+  * For example, consider a notebook with the following cells:
+  *
+  *   cell1:
+  *     x = 1
+  *     y = 2
+  *
+  *   cell2:
+  *     x = 2
+  *
+  *   cell3:
+  *     z = y
+  *
+  *  When we run all the cells (*regardless of which order we ran them*) we want to get a structure that looks like (
+  *  simplifying because we keep a little more info):
+  *
+  *   [
+  *     (cell1, (x -> 1, y -> 2)),
+  *     (cell2, (x -> 2),
+  *     (cell3, (z -> 2)
+  *   ]
+  *
+  *   The order is important, because it determines a relationship between the cells that ensures that the kernel should
+  *   always work as if all cells were run in order, even though users may run cells out-of-order.
+  *
+  *   Imagine that now a user adds a new cell in between cell1 and cell2:
+  *
+  *   cell1:
+  *     x = 1
+  *     y = 2
+  *
+  *   cell4:
+  *     println(z)
+  *
+  *   cell2:
+  *     x = 2
+  *
+  *   cell3:
+  *     z = y
+  *
+  *   Running cell4 should *fail*, saying that `z` is an unknown symbol (even though the same situation in Jupyter or
+  *   Zeppelin would work), because the RuntimeContextView for cell4 would just be [(cell1, (cell1, (x -> 1, y -> 2)))].
+  *   This ensures that notebooks will always work properly when shared with others.
+  */
+class RuntimeContext(val globalInfo: GlobalInfo) extends Serializable {
+
+  case class RuntimeContextEntry(
+    cellId: String,
+    previousCellId: Option[String],
+    symbols: Map[String, globalInfo.RuntimeValue],
+    interpreterContext: Option[InterpreterContext],
+    maybeResult: Option[globalInfo.RuntimeValue]
+  )
+
+  class RuntimeContextView(val parentEntry: RuntimeContextEntry) {
```

Comment:
It seems like `RuntimeContextView` ought to be an unnecessary layer... maybe what's currently `RuntimeContext` belongs in `GlobalInfo` (for sharing `Global` namespace) and `RuntimeContextEntry` can be `RuntimeContext`? Let's discuss

Reply:
Per our discussion I tried my hand at removing the view, I think it looks a bit better now :)